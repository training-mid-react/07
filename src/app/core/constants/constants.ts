
export const ROWS = 6;
export const COLS = 7;
export const WINNING_LENGTH = 4;

export const PLAYER1 = {
    id: 1,
    name: "Player 1",
    color: "#fee813",
};
export const PLAYER2 = {
    id: 2,
    name: "Player 2",
    color: "#fe4c13",
};

