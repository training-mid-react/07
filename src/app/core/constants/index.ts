export const ROWS = 6;
export const COLS = 7;

export const BASE_MATRIZ = new Array(ROWS + 1)
    .fill(null)
    .map(() => new Array(COLS).fill(null));
