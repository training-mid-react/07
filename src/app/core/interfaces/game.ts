export interface GameAction {
  type: string;
  payload?: React.ReactNode;
}
