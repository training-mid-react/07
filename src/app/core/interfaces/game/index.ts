export * from "./boardGame.interface"
export * from "./player.interface"