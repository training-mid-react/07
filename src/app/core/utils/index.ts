export * from './checkWinner';
export * from './checkFourInLine';
export * from './checkTie';
