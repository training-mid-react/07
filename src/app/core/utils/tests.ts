export function rgbaToHex(rgba:string) {
    // Extraer los valores de rgba usando una expresiÃ³n regular
    const [_, r, g, b] = rgba.match(/rgba?\((\d+), (\d+), (\d+)/) || [];

    // Convertir los valores RGB a hexadecimal
    const hex = (c:string) => Number(c).toString(16).padStart(2, "0");
    return `#${hex(r)}${hex(g)}${hex(b)}`;
}