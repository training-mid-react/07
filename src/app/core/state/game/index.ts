import { COLS, PLAYER1, ROWS } from "../../constants/constants";
import { CELL, IGameState } from "../../interfaces/game.interface";
import { IPlayer } from "../../interfaces/player.interface";
import { gameActions } from "./actions";

export const gameInitialState: IGameState = {
    grid: Array(ROWS).fill(Array(COLS).fill(null)),
    playerActive: PLAYER1,
    winner: null,
    isDraw: false,
};

export const gameCases = {
    [gameActions.CHANGE_PLAYER]: (state: IGameState, payload: IPlayer) => {
        return {
            ...state,
            playerActive: payload,
        };
    },
    [gameActions.SET_WINNER]: (state: IGameState, payload: IPlayer) => {
        return {
            ...state,
            winner: payload,
        };
    },
    [gameActions.CHANGE_GRID]: (state: IGameState, payload: CELL[][]) => {
        return {
            ...state,
            grid: payload,
        };
    },
    [gameActions.SET_DRAW]: (state: IGameState, payload: boolean) => {
        return {
            ...state,
            isDraw: payload,
        };
    },
    [gameActions.RESET_GAME]: (state: IGameState, payload: IGameState) => {
        return {
            ...payload,
        };
    },
}