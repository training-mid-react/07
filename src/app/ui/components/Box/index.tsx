import { CELL } from "@/app/core/interfaces/game.interface";
import "./style.scss"

interface Props {
    value: CELL;
    onClick: () => void;
}

const Box = ({ value, onClick }: Props) => {
    return (
        <div data-testid="box" className="box" onClick={onClick}>
            {
                value && <div style={{backgroundColor: value.color}} className="box__chip"></div>
            }
        </div>
    );
}

export default Box;