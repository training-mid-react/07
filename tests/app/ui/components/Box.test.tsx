import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { vi } from "vitest";
import { PLAYER1 } from "../../../../src/app/core/constants/constants";
import { CELL } from '../../../../src/app/core/interfaces/game.interface';
import { rgbaToHex } from '../../../../src/app/core/utils/tests';
import Box from "../../../../src/app/ui/components/Box";


describe("Test de Box", () => {
    // const grid = Array(ROWS).fill(Array(COLS).fill(null))
    const mockHandleClick = vi.fn();
    const value: CELL = null;

    test("Snapshot del Box", () => {
        const { asFragment } = render(<Box value={value} onClick={mockHandleClick} />)
        expect(asFragment).toMatchSnapshot();
    })
    test("Renderiza el Box correctamente", () => {
        const { container } = render(<Box value={value} onClick={mockHandleClick} />)
        expect(container.querySelectorAll(".box").length).toBe(1);
    })
    test("Renderiza el Box con el color correcto", () => {
        const value: CELL = PLAYER1
        const { container } = render(<Box value={value} onClick={mockHandleClick} />)
        const box = container.querySelector(".box__chip");
        const styles = window.getComputedStyle(box!);
        const color = rgbaToHex(styles.backgroundColor);

        expect(color).toBe(value.color);
    })
    test("Llama al mÃ©todo de click del Box", () => {
        const value: CELL = PLAYER1;
        render(<Box value={value} onClick={mockHandleClick} />)
        const box = screen.getByTestId('box')
        fireEvent.click(box!);
        expect(mockHandleClick).toHaveBeenCalledTimes(1);
    })
})
