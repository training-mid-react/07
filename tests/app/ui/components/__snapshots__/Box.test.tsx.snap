// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test de Box > Snapshot del Box 1`] = `[Function]`;
